lib/src/git_backend.rs

```diff
@@ -1002,7 +1007,17 @@ impl Backend for GitBackend {
                 extra_headers: Default::default(),
             };
 
-            // todo sign commits here
+            if let Some(sign) = &mut sign_with {
+                // we don't use gix pool, but at least use their heuristic
+                let mut data = Vec::with_capacity(512);
+                commit.write_to(&mut data).unwrap();
+
+                let sig = sign(&data)?;
+                commit
+                    .extra_headers
+                    .push(("gpgsig".into(), sig.clone().into()));
```

Comment:
> the backend may need to include e.g. commit headers saying which format was used

Hm, right, I see
I guess it does not hurt to just have a "callback with metadata attached" instead of just a function here, and metadata could be a enum, or maybe a signer implementation provided ('static) string or something;
Also I've noticed that several times I want to go and do some associated type stuff, but cant because we have trait objects instead of heaps of generics in generics in generics :) 

Heh and yeah in the end I just settled on your initial idea of passing a signing callback to the write_commit ¯\\\_(ツ)_/¯

Reply:
Hmm, actually, looking in the [future](https://github.com/martinvonz/jj/pull/1738/files#diff-f189927fcb04c403e436400333c7eeda417a658848c875422fd8ef7382efb38eR172-R174), there will be a need to determine the signature type (that is, is it pgp or ssh or x509 or whatever else) from the signature alone.

In your example, "what if the backend needs to store the signature type separately in another header)", it makes sense for the backend to then also give this type back to the signer backend methods - but in git, for example, there will have to be this gross "determine the signature type from the signature alone since git doesn't store that".

My conclusion in the end is that I probably should undo this last change and just have it be a function. Pretty sure that for all known git signature types (mentioned pgp, ssh and x509) the signature type can be determined just from the armored prefix anyway.

Hope you understood my rambling here :)

Reply:
I think I follow, but how do we know here that `sign` should use the `gpgsig` and not some other header (I don't know what other options there are)?

Reply:
`gpgsig` is just a historical name 🤷
Any signatures are just stored in it, and git determines what "signing backend" to use based on the armored prefixes just like I'm planning to do too.

There's also the new `gpgsig-sha256` related to https://git-scm.com/docs/hash-function-transition, but it didn't really lift off so far I think?. We ~~do read it~~ [support](https://github.com/martinvonz/jj/pull/2557/files#diff-0a6cf0d1c0a8ecc62981eab5d4fe0c6ce9f1802e00a86eb6278836e031e337d2R425) reading it once gix does, and it doesn't currently; and updating to write it should be trivial in the future

Reply:
> `gpgsig` is just a historical name 🤷

Ohh, I see :) Just a function seems appropriate then. Thanks.